### ISSUE TITLE
zimbra_milter.py: `self.user` is `NoneType` on any but first connection

### ISSUE THREAD

**Issue by JM-Lemmi** [cirosec-studis/zimbraweb-smtp-bridge#50]
Its probably fine, since the milter still rejects it, but maybe it should be caught to prevent issues?

## Cmd to Reproduce

```powershell
Send-MailMessage -SmtpServer localhost -Port 587 -To s212689@student.dhbw-mannheim.de -From s212689@student.dhbw-mannheim.de -Subject Test
```

## Logs

```
Traceback (most recent call last):
  File "/srv/zimbraweb/zimbra_milter.py", line 162, in <lambda>
    lambda ctx, *b: ctx.getpriv().envfrom_bytes(*b))
  File "/usr/lib/python3.9/site-packages/Milter/__init__.py", line 363, in envfrom_bytes
    return self.envfrom(s[0],*s[1:])
  File "/srv/zimbraweb/zimbra_milter.py", line 59, in envfrom
    if "@" in self.user:
TypeError: argument of type 'NoneType' is not iterable
Dec 23 20:04:40 d5b6e78e39af postfix/submission/smtpd[119]: connect from unknown[172.17.0.1]
Dec 23 20:04:40 d5b6e78e39af postfix/submission/smtpd[119]: NOQUEUE: milter-reject: MAIL from unknown[172.17.0.1]: 451 4.3.0 pymilter: untrapped exception in pythonfilter; from=<s212689@student.dhbw-mannheim.de> proto=ESMTP helo=<Julian-X380>
```

**Comment by JM-Lemmi** [comment]
Nvm, I think I fucked something up in commit 33b027c

**Comment by frereit** [comment]
Ist das sicher closed? Das Traceback lässt mich vermuten, dass es ein Fehler von mir war

**Comment by JM-Lemmi** [comment]
I'm not 100% sure. I'm working on it in #41. My current diagnosis is, that this error appears, when the logger is set to a file.

**Comment by JM-Lemmi** [comment]
This bug is not ironed out. I did think it had to do with logging to a file the milter is not allowed to open and it was gone, but now I'm running into issues again, when sending HTML E-Mails. Sending plain text mails is fine

## Reproduce

Commit 93e23e1

```
smtp_fallback: enabled
```

[docker-compose.yml](https://gist.github.com/JM-Lemmi/7c308b8cd256dfe1417adace51360c0a#file-docker-compose-yml)

Send HTML Mail with following command
```powershell
Send-MailMessage -SmtpServer j-cx11-web.cloud.julian-lemmerich.de -Port 587 -To s212689@student.dhbw-mannheim.de -From s212689@student.dhbw-mannheim.de -Subject Test -BodyAsHtml Test -Credential (Get-Credential)
```

## Logs

```
zimbrabridge_1  | Dec 28 20:21:13 dhbwbridge postfix/submission/smtpd[114]: connect from unknown[46.114.1.221]
zimbrabridge_1  | Dec 28 20:21:14 dhbwbridge python/zimbra_authentication.py: Trying to authenticate user AUTH_USERNAME='s212689'
zimbrabridge_1  | Dec 28 20:21:14 dhbwbridge python/zimbra_authentication.py: successfully authenticated AUTH_USERNAME='s212689'
zimbrabridge_1  | Traceback (most recent call last):
zimbrabridge_1  |   File "/srv/zimbraweb/zimbra_milter.py", line 159, in <lambda>
zimbrabridge_1  |     lambda ctx, *b: ctx.getpriv().envfrom_bytes(*b))
zimbrabridge_1  |   File "/usr/lib/python3.9/site-packages/Milter/__init__.py", line 363, in envfrom_bytes
zimbrabridge_1  |     return self.envfrom(s[0],*s[1:])
zimbrabridge_1  |   File "/srv/zimbraweb/zimbra_milter.py", line 56, in envfrom
zimbrabridge_1  |     if "@" in self.user:
zimbrabridge_1  | TypeError: argument of type 'NoneType' is not iterable
zimbrabridge_1  | Dec 28 20:21:16 dhbwbridge postfix/submission/smtpd[114]: warning: unknown[46.114.1.221]: SASL login authentication failed: UGFzc3dvcmQ6
zimbrabridge_1  | Dec 28 20:21:16 dhbwbridge postfix/submission/smtpd[114]: NOQUEUE: milter-reject: MAIL from unknown[46.114.1.221]: 451 4.3.0 pymilter: untrapped exception in pythonfilter; from=<s212689@student.dhbw-mannheim.de> proto=ESMTP helo=<Julian-X380>
zimbrabridge_1  | Dec 28 20:21:16 dhbwbridge postfix/submission/smtpd[114]: lost connection after MAIL from unknown[46.114.1.221]
zimbrabridge_1  | Dec 28 20:21:16 dhbwbridge postfix/submission/smtpd[114]: disconnect from unknown[46.114.1.221] ehlo=1 auth=0/1 mail=0/1 commands=1/3
```

May be related to #54

**Comment by JM-Lemmi** [comment]
So now it works again. I can't reproduce the error anymore. But it came up a few times now, so I will leave this issue open for now

**Comment by JM-Lemmi** [comment]
Maybe there is a correlation with preexisting config file? I think removing all preexisting volumes from docker compose with `docker-compose rm -v` gets rid of the error for the next start

**Comment by JM-Lemmi** [comment]
I'm pretty sure only the first mail succeeds. After a restart again, one mail can be sent.

**Comment by JM-Lemmi** [comment]
As long as the TLS connection is open and no new authentication is required it works. But waiting for client disconnect and trying again will crash

https://gist.github.com/JM-Lemmi/9094914259f3a7ee9a75fffe9de39ec6

**Comment by JM-Lemmi** [comment]
Can this have something to do with this? Because we have existing users, that are talked about here?

https://github.com/cirosec-studis/zimbraweb-smtp-bridge/blob/8a57c8f5a891ffe75eb7a8949dfa0cef253f5102/files/zimbra_milter.py#L149-L152

**Comment by JM-Lemmi** [comment]
Commit d718bd1 is the first to include the bug (literally the first after v1.1.0

**Comment by JM-Lemmi** [comment]
There is not much difference left now in the Dockerfile. I suspect its something from the Dockerfile since its the only thing changed in d718bd1

[diff between d718bd1 and v1.0](https://github.com/cirosec-studis/zimbraweb-smtp-bridge/compare/v1.0.0..d718bd1)

[diff from patch50 & v1.0](https://github.com/cirosec-studis/zimbraweb-smtp-bridge/compare/v1.0.0..patch50#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557)

**Comment by JM-Lemmi** [comment]
Der Fehler hängt nicht am Nutzer. Der gleiche Nutzer kann sich von zwei verschiedenen Clients einloggen. Jeweils die erste Verbindung funktioniert problemlos, alle weiteren Verbindungen schlagen fehl.

Dies lässt sich reproduzieren, indem man mehrere Powershell-Container startet (diese Simulieren die Clients) und dann hintereinander von diesen Mails verschickt und dazwischen wartet, bis die TCP Verbindung abgelaufen ist.

Tip: Aktive TCP Verbindungen lassen sich mit [Sysinternals TCPview](https://docs.microsoft.com/en-us/sysinternals/downloads/tcpview) gut ansehen.

**Comment by JM-Lemmi** [comment]
> Commit [d718bd1](https://github.com/cirosec-studis/zimbraweb-smtp-bridge/commit/d718bd13282e141e568d243435b84212d627bcc3) is the first to include the bug (literally the first after v1.1.0)

I was wrong, actually 6234674 is the first commit after v1.0.0 to include this bug

**Comment by JM-Lemmi** [comment]
Reverting the changes from #31 fixed it. I will close this issue and reopen it.

**Comment by frereit** [comment]
Yup, sorry about the delay on my part. 

The issue is that the `if "@" in self.user` check is done before the `if self.user:` check, so when no user is set, we get an error. I will deploy a fix and then v1.1 should be ready
